spring-pulsar/src/main/java/org/springframework/pulsar/config/GenericPulsarListenerEndpointRegistry.java

```diff
@@ -0,0 +1,265 @@
+ *
+ * @param <C> listener container type.
+ * @param <E> listener endpoint type.
+ * @author Soby Chacko
+ * @author Christophe Bornet
+ */
+public class GenericPulsarListenerEndpointRegistry<C extends SmartLifecycle & DisposableBean, E extends GenericPulsarListenerEndpoint>
+		implements GenericPulsarListenerContainerRegistry<C>, DisposableBean, SmartLifecycle, ApplicationContextAware,
+		ApplicationListener<ContextRefreshedEvent> {
+
+	private final Class<? extends C> type;
+
+	private final Map<String, C> listenerContainers = new ConcurrentHashMap<>();
+
+	private ConfigurableApplicationContext applicationContext;
+
+	private int phase = C.DEFAULT_PHASE;
+
+	private boolean contextRefreshed;
+
+	private volatile boolean running;
+
+	@SuppressWarnings("unchecked")
+	protected GenericPulsarListenerEndpointRegistry(Class<?> type) {
+		this.type = (Class<? extends C>) type;
+	}
+
+	@Override
+	public void setApplicationContext(ApplicationContext applicationContext) throws BeansException {
+		if (applicationContext instanceof ConfigurableApplicationContext) {
+			this.applicationContext = (ConfigurableApplicationContext) applicationContext;
+		}
+	}
+
+	@Override
+	@Nullable
+	public C getListenerContainer(String id) {
+		Assert.hasText(id, "Container identifier must not be empty");
+		return this.listenerContainers.get(id);
+	}
+
+	@Override
+	public Set<String> getListenerContainerIds() {
+		return Collections.unmodifiableSet(this.listenerContainers.keySet());
+	}
+
+	@Override
+	public Collection<C> getListenerContainers() {
+		return Collections.unmodifiableCollection(this.listenerContainers.values());
+	}
+
+	@Override
+	public Collection<C> getAllListenerContainers() {
+		List<C> containers = new ArrayList<>(getListenerContainers());
+		containers.addAll(this.applicationContext.getBeansOfType(this.type, true, false).values());
+		return containers;
+	}
+
+	public void registerListenerContainer(E endpoint, GenericPulsarListenerContainerFactory<? extends C, E> factory) {
+		registerListenerContainer(endpoint, factory, false);
+	}
+
+	public void registerListenerContainer(E endpoint, GenericPulsarListenerContainerFactory<? extends C, E> factory,
+			boolean startImmediately) {
+		Assert.notNull(endpoint, "Endpoint must not be null");
+		Assert.notNull(factory, "Factory must not be null");
+
+		String subscriptionName = endpoint.getSubscriptionName();
+		String id = endpoint.getId();
+
+		Assert.hasText(subscriptionName, "Endpoint id must not be empty");
+
+		synchronized (this.listenerContainers) {
+			Assert.state(!this.listenerContainers.containsKey(id),
+					"Another endpoint is already registered with id '" + subscriptionName + "'");
+			C container = createListenerContainer(endpoint, factory);
+			this.listenerContainers.put(id, container);
+		}
+	}
+
+	protected C createListenerContainer(E endpoint, GenericPulsarListenerContainerFactory<? extends C, E> factory) {
+
+		if (endpoint instanceof MethodPulsarListenerEndpoint<?> mkle) {
```

Comment:
Oups. Something to do here

Reply:
Actually this seems to be dead code as the only call  of `EndpointHandlerMethod`'s constructor is inside this check. Probably some leftover from spring-kafka.
I propose to remove `EndpointHandlerMethod`. OK with that ?

Reply:
Yeh, this section of the code is doing quite a bit of things and when we find the dead code that was used in `spring-kafka` but not here in `spring-pulsar` too am in favor of pulling. 

That being said, I am also in favor of changing as little as "live" imperative code in this particular PR as its already big. I would much rather see a list of todos for a small subsequent PR. Not talking directly about this change, but just in general. I am about to review the rest of the PR now.

Reply:
Ok